[Bug] Panic (nil pointer) in metaBuilder.addSegments when collectionBackups has no entry for segment.GetCollectionId()

### Milvus-Backup Version

v0.5.12

### Can you reproduce this issue on the latest version?

Yes, it also happens on the latest version

### Current Behavior

[2026/04/01 02:05:44.961 +08:00] [INFO] [backup/coll_dml_task.go:302] ["try get segment info via proxy node"] [task_id=1301615a-35dc-4b7f-b5e8-9cd50738732d] [ns=default.code_66193035]       
[segment_id=465136321155628517]                                                                                                                                                               
panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                       
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x1b5205e]                                                                                                                      
                                                                                                                                                                                              
goroutine 36725 [running]:                                                                                                                                                                    
github.com/zilliztech/milvus-backup/core/backup.(*metaBuilder).addSegments(0xc000fd4c30, {0xc00382b380, 0x1, 0xc000428830?})                                                                  
        /home/runner/work/milvus-backup/milvus-backup/core/backup/meta_builder.go:101 +0x13e                                                                                                  
github.com/zilliztech/milvus-backup/core/backup.(*collDMLTask).Execute(0xc00306cd20, {0x2d9a460, 0xc001eea190})                                                                               
        /home/runner/work/milvus-backup/milvus-backup/core/backup/coll_dml_task.go:359 +0x294                                                                                                 
github.com/zilliztech/milvus-backup/core/backup.(*serialFlushStrategy).executeDMLTask.func1({0x2d9a460, 0xc001eea190})                                                                        
        /home/runner/work/milvus-backup/milvus-backup/core/backup/coll_strategy.go:242 +0x24d                                                                                                 
github.com/zilliztech/milvus-backup/core/backup.concurrentExecCollTask.func1()                                                                                                                
        /home/runner/work/milvus-backup/milvus-backup/core/backup/coll_strategy.go:78 +0x5c                                                                                                   
golang.org/x/sync/errgroup.(*Group).Go.func1()                                                                                                                                                
        /home/runner/go/pkg/mod/golang.org/x/sync@v0.19.0/errgroup/errgroup.go:93 +0x50                                                                                                       
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 1                                                                                                                              
        /home/runner/go/pkg/mod/golang.org/x/sync@v0.19.0/errgroup/errgroup.go:78 +0x95                                                                                                       
[2026-04-01 02:05:44] 错误: milvus-backup create 失败:

### Expected Behavior

- If collectionBackups[segment.GetCollectionId()] is missing, do not panic.
- Log a structured error with collection_id, segment_id, and optionally db/collection name, and fail the backup task with a readable error so operators can investigate Milvus metadata or version mismatch

### Steps To Reproduce

```markdown

```

### Environment

```markdown
milvus-back-up:v0.5.12
milvus:2.6.4
```

### Anything else?

**Root cause (analysis):**
In addSegments, the code does:
`collBackup := builder.collectionBackups[segment.GetCollectionId()]
collBackup.Size += segment.GetSize()  // panics if collBackup is nil`

collectionBackups is populated during the DDL phase from DescribeCollection. Segment CollectionId comes from PersistentSegmentInfo (and the same segment is used when building SegmentBackupInfo in getSegmentInfoByAPI). If those IDs diverge or CollectionId is 0 / invalid, the map lookup returns nil and the next line dereferences it.
There is already a warning + continue path when a partition is missing (partitionBackups); collection lookup has no equivalent guard, which makes this failure mode a hard crash.

**Suggested fix / improvement:**
1. After resolving collBackup, if collBackup == nil, log Warn/Error with segment.GetSegmentId(), 2. segment.GetCollectionId(), and fail fast with fmt.Errorf(...).
2. Optionally validate segment != nil and non-zero CollectionId before appending to bakSegs in getSegments / getSegment.
3. Add a small unit test for addSegments with an unknown CollectionId to assert error instead of panic.

--- a/core/backup/coll_dml_task.go
+++ b/core/backup/coll_dml_task.go
@@ -356,7 +356,9 @@ func (dmlt *collDMLTask) Execute(ctx context.Context) error {
 		return fmt.Errorf("backup: get segments %w", err)
 	}
 
-	dmlt.metaBuilder.addSegments(segments)
+	if err := dmlt.metaBuilder.addSegments(segments); err != nil {
+		return fmt.Errorf("backup: add segments meta: %w", err)
+	}
 
 	size := lo.SumBy(segments, func(seg *backuppb.SegmentBackupInfo) int64 { return seg.GetSize() })
 	dmlt.taskMgr.UpdateBackupTask(dmlt.taskID, taskmgr.SetBackupCollDMLExecuting(dmlt.ns, size))

--- a/core/backup/coll_strategy.go
+++ b/core/backup/coll_strategy.go
@@ -216,7 +216,9 @@ func (sf *serialFlushStrategy) flushAndBackupPOS(ctx context.Context, ns namespa
 		maxChannelTS = max(maxChannelTS, checkpoint.GetTimestamp())
 	}
 
-	sf.args.MetaBuilder.addPOS(ns, channelCP, maxChannelTS, uint64(resp.GetCollSealTimes()[ns.CollName()]))
+	if err := sf.args.MetaBuilder.addPOS(ns, channelCP, maxChannelTS, uint64(resp.GetCollSealTimes()[ns.CollName()])); err != nil {
+		return fmt.Errorf("backup: add POS meta: %w", err)
+	}
 	return nil
 }
 

--- a/core/backup/meta_builder.go
+++ b/core/backup/meta_builder.go
@@ -78,26 +78,35 @@ func (builder *metaBuilder) addCollection(ns namespace.NS, collectionBackup *bac
 	}
 }
 
-func (builder *metaBuilder) addPOS(ns namespace.NS, channelCP map[string]string, maxChannelTS uint64, sealTime uint64) {
+func (builder *metaBuilder) addPOS(ns namespace.NS, channelCP map[string]string, maxChannelTS uint64, sealTime uint64) error {
 	builder.mu.Lock()
 	defer builder.mu.Unlock()
 
 	collID := builder.nsToCollID[ns]
-	collBackup := builder.collectionBackups[collID]
+	collBackup, ok := builder.collectionBackups[collID]
+	if !ok {
+		return fmt.Errorf("backup: collection backup not found for namespace %s, collection_id: %d", ns, collID)
+	}
 
 	collBackup.ChannelCheckpoints = channelCP
 	collBackup.BackupTimestamp = maxChannelTS
 	collBackup.BackupPhysicalTimestamp = sealTime
+
+	return nil
 }
 
-func (builder *metaBuilder) addSegments(segments []*backuppb.SegmentBackupInfo) {
+func (builder *metaBuilder) addSegments(segments []*backuppb.SegmentBackupInfo) error {
 	builder.mu.Lock()
 	defer builder.mu.Unlock()
 
 	for _, segment := range segments {
-		builder.data.Size += segment.GetSize()
+		collBackup, ok := builder.collectionBackups[segment.GetCollectionId()]
+		if !ok {
+			return fmt.Errorf("backup: collection backup not found for segment %d, collection_id: %d",
+				segment.GetSegmentId(), segment.GetCollectionId())
+		}
 
-		collBackup := builder.collectionBackups[segment.GetCollectionId()]
+		builder.data.Size += segment.GetSize()
 		collBackup.Size += segment.GetSize()
 
 		if segment.GetIsL0() && segment.GetPartitionId() == _allPartitionID {
@@ -117,6 +126,8 @@ func (builder *metaBuilder) addSegments(segments []*backuppb.SegmentBackupInfo) 
 			partBackup.Size += segment.GetSize()
 		}
 	}
+
+	return nil
 }
 
 func (builder *metaBuilder) addIndexExtraInfo(indexes []*indexpb.FieldIndex) {

--- a/core/backup/meta_builder_test.go
+++ b/core/backup/meta_builder_test.go
@@ -10,7 +10,7 @@ import (
 	"github.com/zilliztech/milvus-backup/internal/namespace"
 )
 
-func newTestMetaBuilder() *metaBuilder {
+func newTestMetaBuilder(t *testing.T) *metaBuilder {
 	builder := newMetaBuilder("task1", "backup1")
 
 	ns := namespace.New("db1", "coll1")
@@ -60,13 +60,44 @@ func newTestMetaBuilder() *metaBuilder {
 			},
 		},
 	}
-	builder.addSegments(append(segments, l0Segments...))
+	assert.NoError(t, builder.addSegments(append(segments, l0Segments...)))
 
 	return builder
 }
 
+func TestAddSegments(t *testing.T) {
+	t.Run("UnknownCollectionID", func(t *testing.T) {
+		builder := newMetaBuilder("task1", "backup1")
+
+		segments := []*backuppb.SegmentBackupInfo{
+			{
+				SegmentId:    100,
+				CollectionId: 999,
+				PartitionId:  10,
+				Size:         4096,
+			},
+		}
+
+		err := builder.addSegments(segments)
+		assert.Error(t, err)
+		assert.Contains(t, err.Error(), "backup: collection backup not found")
+		assert.Contains(t, err.Error(), "999")
+	})
+}
+
+func TestAddPOS(t *testing.T) {
+	t.Run("UnknownNamespace", func(t *testing.T) {
+		builder := newMetaBuilder("task1", "backup1")
+
+		ns := namespace.New("db_unknown", "coll_unknown")
+		err := builder.addPOS(ns, map[string]string{"ch1": "cp1"}, 100, 200)
+		assert.Error(t, err)
+		assert.Contains(t, err.Error(), "backup: collection backup not found")
+	})
+}
+
 func TestBuildCollectionMetaStripsSegments(t *testing.T) {
-	builder := newTestMetaBuilder()
+	builder := newTestMetaBuilder(t)
 
 	data, err := builder.buildCollectionMeta()
 	assert.NoError(t, err)
@@ -88,7 +119,7 @@ func TestBuildCollectionMetaStripsSegments(t *testing.T) {
 }
 
 func TestBuildPartitionMetaStripsSegments(t *testing.T) {
-	builder := newTestMetaBuilder()
+	builder := newTestMetaBuilder(t)
 
 	data, err := builder.buildPartitionMeta()
 	assert.NoError(t, err)
@@ -105,7 +136,7 @@ func TestBuildPartitionMetaStripsSegments(t *testing.T) {
 }
 
 func TestBuildSegmentMetaContainsAllSegments(t *testing.T) {
-	builder := newTestMetaBuilder()
+	builder := newTestMetaBuilder(t)
 
 	data, err := builder.buildSegmentMeta()
 	assert.NoError(t, err)
@@ -124,7 +155,7 @@ func TestBuildSegmentMetaContainsAllSegments(t *testing.T) {
 }
 
 func TestBuildFullMetaContainsEverything(t *testing.T) {
-	builder := newTestMetaBuilder()
+	builder := newTestMetaBuilder(t)
 
 	data, err := builder.buildFullMeta()
 	assert.NoError(t, err)
@@ -146,7 +177,7 @@ func TestBuildFullMetaContainsEverything(t *testing.T) {
 }
 
 func TestSequentialBuildDoesNotCorruptData(t *testing.T) {
-	builder := newTestMetaBuilder()
+	builder := newTestMetaBuilder(t)
 
 	// Simulate production call order: collection → partition → segment → full
 	_, err := builder.buildCollectionMeta()

